File: python/dolfin/constant.py
# Copyright (C) 2019 Michal Habera and Chris Richardson
#
# This file is part of DOLFIN (https://www.fenicsproject.org)
#
# SPDX-License-Identifier:    LGPL-3.0-or-later

from typing import Sequence, Union

import numpy

import dolfin.cpp
import ufl


class Constant(ufl.Constant):
    def __init__(self, domain, c: Union[numpy.ndarray, Sequence, float]):
        """A constant with respect to a domain.

        Parameters
        ----------
        domain : DOLFIN or UFL mesh
        c
            Value of the constant.
        """
        c_np = numpy.asarray(c)
        super().__init__(domain, c_np.shape)
        self._cpp_object = dolfin.cpp.function.Constant(c_np.shape, c_np.flatten())

    @property
    def value(self):
        """Returns value of the constant."""
        return self._cpp_object.value()

    @value.setter
    def value(self, v):
        numpy.copyto(self._cpp_object.value(), numpy.asarray(v))

File: cpp/dolfinx/function/Constant.cpp
// Copyright (C) 2019 Chris Richardson and Michal Habera
//
// This file is part of DOLFINX (https://www.fenicsproject.org)
//
// SPDX-License-Identifier:    LGPL-3.0-or-later

#include "Constant.h"
#include <Eigen/Dense>
#include <vector>

using namespace dolfinx;
using namespace dolfinx::function;

//-----------------------------------------------------------------------------
Constant::Constant(PetscScalar c) : value({c})
{
  // Do nothing
}
//-----------------------------------------------------------------------------
Constant::Constant(const std::vector<PetscScalar>& c)
    : shape(1, c.size()), value({c})
{
  // Do nothing
}
//-----------------------------------------------------------------------------
Constant::Constant(
    const Eigen::Ref<Eigen::Array<PetscScalar, Eigen::Dynamic, Eigen::Dynamic,
                                  Eigen::RowMajor>>& c)
    : shape({(int)c.rows(), (int)c.cols()}), value(c.rows() * c.cols())
{
  // Copy data from Eigen::Array to flattened vector
  for (int i = 0; i < c.rows(); ++i)
    for (int j = 0; j < c.cols(); ++j)
      value[i * c.cols() + j] = c(i, j);
}
//-----------------------------------------------------------------------------
Constant::Constant(std::vector<int> shape, std::vector<PetscScalar> c)
    : shape(shape), value(c)
{
  // Do nothing
}
//-----------------------------------------------------------------------------
